lls/arp.c

```diff
@@ -0,0 +1,541 @@
+	return 0;
+}
+
+/*
+ * Returns 0 if the packet needs to be transmitted,
+ * -1 if it does not need to be transmitted (and can be freed).
+ */
+int
+process_arp(struct gatekeeper_if *iface, struct ether_hdr *eth_hdr,
+	uint16_t off)
+{
+	struct arp_hdr *arp_hdr =
+		(struct arp_hdr *)((char *)(eth_hdr + 1) + off);
+	uint32_t target_ip_be_32;
+
+	/* We only support ARP for resolving IPv4 --> Ethernet addresses. */
+	if (unlikely(arp_hdr->arp_hrd != rte_cpu_to_be_16(ARP_HRD_ETHER) ||
+		     arp_hdr->arp_pro != rte_cpu_to_be_16(ETHER_TYPE_IPv4) ||
+		     arp_hdr->arp_hln != ETHER_ADDR_LEN ||
+		     arp_hdr->arp_pln != sizeof(uint32_t)))
+		return -1;
+
+	/* XXX If sip is not in the same subnet as our IP address, drop. */
+
+	/* Check if we have this IP address. */
+	target_ip_be_32 = arp_hdr->arp_data.arp_tip;
+	if (iface->ipv4_addr != rte_be_to_cpu_32(target_ip_be_32))
+		return -1;
+
+	/* Update cache with source resolution, regardless of operation. */
+	arp_cache_add(rte_be_to_cpu_32(arp_hdr->arp_data.arp_sip),
+		&arp_hdr->arp_data.arp_sha, iface->id);
+
+	switch (rte_be_to_cpu_16(arp_hdr->arp_op)) {
+	case ARP_OP_REQUEST: {
+		/* Set-up Ethernet header. */
+		ether_addr_copy(&eth_hdr->s_addr, &eth_hdr->d_addr);
+		ether_addr_copy(&iface->eth_addr, &eth_hdr->s_addr);
+
+		/* Set-up ARP header. */
+		arp_hdr->arp_op = rte_cpu_to_be_16(ARP_OP_REPLY);
+		ether_addr_copy(&arp_hdr->arp_data.arp_sha,
+			&arp_hdr->arp_data.arp_tha);
+		arp_hdr->arp_data.arp_tip = arp_hdr->arp_data.arp_sip;
+		ether_addr_copy(&iface->eth_addr, &arp_hdr->arp_data.arp_sha);
+		arp_hdr->arp_data.arp_sip = target_ip_be_32;
+
+		/* Need to transmit reply. */
+		return 0;
+	}
+	case ARP_OP_REPLY:
+		/*
+		 * No further action required. Could check to make sure
+		 * arp_hdr->arp_data.arp_tha is equal to lls->ether_addr,
+		 * but there's nothing that can be done if it's wrong anyway.
+		 */
+		return -1;
+	default:
+		return -1;
+	}
+
+	RTE_ASSERT(0);
+}
+
+void
+arp_cache_destroy(void)
+{
+	rte_timer_stop(&arp_cache_timer);
+	rte_ring_free(arp_cache_ring);
+	rte_mempool_free(arp_cache_entry_pool);
+	rte_mempool_free(arp_cache_req_pool);
+	rte_hash_free(arp_cache);
+}
+
+int
+arp_cache_init(void)
+{
+	/* Finalize parameters and create hash. */
+	arp_cache_params.socket_id = rte_socket_id();
+	arp_cache = rte_hash_create(&arp_cache_params);
+	if (arp_cache == NULL) {
+		RTE_LOG(ERR, HASH, "Could not create ARP cache hash\n");
+		return -1;
+	}
+
+	/* Allocate pools for entries and requests to change cache. */
+	arp_cache_req_pool = rte_mempool_create("arp_cache_req_pool",
+		ARP_CACHE_REQS, sizeof(struct arp_cache_req),
+		ARP_CACHE_REQS_MEMPOOL_SIZE, 0, NULL, NULL, NULL, NULL,
+		rte_socket_id(), 0);
+	if (arp_cache_req_pool == NULL) {
+		RTE_LOG(ERR, MEMPOOL, "Could not create ARP request pool\n");
+		goto hash;
+	}
+
+	arp_cache_entry_pool = rte_mempool_create("arp_cache_entry_pool",
+		ARP_CACHE_ENTRIES / 2 - 1, sizeof(struct arp_cache_entry),
```

Comment:
Should the size of the memory pool for entries be at least greater than ``ARP_CACHE_ENTRIES``? Since all the entries in the array should be from the memory pool.